provisioning/device/src/tpm_registration.ts

```diff
@@ -0,0 +1,239 @@
+            - `sasToken`: the SAS token generated according to `SRS_NODE_DPS_TPM_REGISTRATION_16_006`
+            - `registrationInfo`: an object with the following properties `endorsementKey`, `storageRootKey`, `registrationId` and their previously set values.
+            - a callback that will handle an optional error and a `result` object containing the IoT hub name, device id and symmetric key for this device.]*/
+            this._transport.register(registrationInfo, sasToken, (err, result) => {
+              if (err) {
+                debug('failed to register with provisioning transport: ' + err.toString());
+                /*Codes_SRS_NODE_DPS_TPM_REGISTRATION_16_010: [If any of the calls the the `TpmSecurityClient` or the `TpmProvisioningTransport` fails, the `register` method shall call its callback with the error resulting from the failure.]*/
+                this._fsm.transition('notStarted', err, registerCallback);
+              } else {
+                this._fsm.transition('storingSecret', result, registerCallback);
+              }
+            });
+          },
+          cancel: (callback) => {
+            /*Codes_SRS_NODE_DPS_TPM_REGISTRATION_16_011: [The `cancel` method shall interrupt the ongoing registration process.]*/
+            this._transport.cancel((err) => {
+              if (err) {
+                debug('failed to stop provisioning transport: ' + err.toString());
+              }
+              this._fsm.transition('notStarted', err, callback);
+            });
+          },
+          '*': () => this._fsm.deferUntilTransition()
+        },
+        storingSecret: {
+          _onEnter: (registrationResult, registerCallback) => {
+            /*Codes_SRS_NODE_DPS_TPM_REGISTRATION_16_008: [When the callback for the registration process is called, the `register` method shall store the symmetric key within the TPM by calling the `activateIdentityKey` method of the `TpmSecurityClient` object passed to the constructor with the following arguments:
+            - `symmetricKey`: the symmetric key property of the `TpmChallenge` object returned by the previous call to `TpmProvisioningTransport.getAuthenticationChallenge`
+            - a callback that will handle an optional error if the operation fails.
+            ]*/
+            this._securityClient.activateIdentityKey(registrationResult.symmetricKey, (err) => {
+              if (err) {
+                debug('failed to stop provisioning transport: ' + err.toString());
+                /*Codes_SRS_NODE_DPS_TPM_REGISTRATION_16_010: [If any of the calls the the `TpmSecurityClient` or the `TpmProvisioningTransport` fails, the `register` method shall call its callback with the error resulting from the failure.]*/
+                this._fsm.transition('notStarted', err, registerCallback);
+              } else {
+                this._fsm.transition('completed', registrationResult, registerCallback);
```

Comment:
>this._fsm.transition('completed', registrationResult, registerCallback); [](start = 16, length = 72)

where do we close the TLS connection?  I know we don't expose the endSession operation to the caller, but what level is responsible for closing the connection?  (I'm thinking the transport could do this internally, but we should make sure we agree)

Reply:
I think the transport should take care of it since the connection is just an artifact of the transport handling the registration request. 

---
In reply to: [152069440](https://github.com/Azure/azure-iot-sdk-node/pull/155#discussion_r152069440) [](ancestors = 152069440)